substrate/client/rpc-spec-v2/src/transaction/transaction_broadcast.rs

```diff
@@ -0,0 +1,217 @@
+	handle: AbortHandle,
+}
+
+impl<Pool, Client> TransactionBroadcast<Pool, Client> {
+	/// Creates a new [`TransactionBroadcast`].
+	pub fn new(client: Arc<Client>, pool: Arc<Pool>, executor: SubscriptionTaskExecutor) -> Self {
+		TransactionBroadcast { client, pool, executor, broadcast_ids: Default::default() }
+	}
+
+	/// Generate an unique operation ID for the `transaction_broadcast` RPC method.
+	pub fn generate_unique_id(&self) -> String {
+		let generate_operation_id = || {
+			// The lenght of the operation ID.
+			const OPERATION_ID_LEN: usize = 16;
+
+			rand::thread_rng()
+				.sample_iter(Alphanumeric)
+				.take(OPERATION_ID_LEN)
+				.map(char::from)
+				.collect::<String>()
+		};
+
+		let mut id = generate_operation_id();
+
+		let broadcast_ids = self.broadcast_ids.read();
+
+		while broadcast_ids.contains_key(&id) {
+			id = generate_operation_id();
+		}
+
+		id
+	}
+}
+
+/// Currently we treat all RPC transactions as externals.
+///
+/// Possibly in the future we could allow opt-in for special treatment
+/// of such transactions, so that the block authors can inject
+/// some unique transactions via RPC and have them included in the pool.
+const TX_SOURCE: TransactionSource = TransactionSource::External;
+
+#[async_trait]
+impl<Pool, Client> TransactionBroadcastApiServer<BlockHash<Pool>>
+	for TransactionBroadcast<Pool, Client>
+where
+	Pool: TransactionPool + Sync + Send + 'static,
+	Pool::Hash: Unpin,
+	<Pool::Block as BlockT>::Hash: Unpin,
+	Client: HeaderBackend<Pool::Block> + BlockchainEvents<Pool::Block> + Send + Sync + 'static,
+{
+	fn broadcast(&self, bytes: Bytes) -> RpcResult<Option<String>> {
+		let pool = self.pool.clone();
+
+		// The unique ID of this operation.
+		let id = self.generate_unique_id();
+
+		let mut best_block_import_stream =
+			Box::pin(self.client.import_notification_stream().filter_map(
+				|notification| async move { notification.is_new_best.then_some(notification.hash) },
+			));
+
+		let broadcast_transaction_fut = async move {
+			// There is nothing we could do with an extrinsic of invalid format.
+			let Ok(decoded_extrinsic) = TransactionFor::<Pool>::decode(&mut &bytes[..]) else {
+				return
+			};
+
+			// Flag to determine if the we should broadcast the transaction again.
+			let mut is_done = false;
+
+			while !is_done {
+				// Wait for the next block to become available.
+				let Some(mut best_block_hash) = best_block_import_stream.next().await else {
+					return
+				};
+				// We are effectively polling the stream for the last available item at this time.
+				// The `now_or_never` returns `None` if the stream is `Pending`.
+				//
+				// If the stream contains `Hash0x1 Hash0x2 Hash0x3 Hash0x4`, we want only `Hash0x4`.
+				while let Some(next) = best_block_import_stream.next().now_or_never() {
+					let Some(next) = next else {
+						// Nothing to do if the best block stream terminated.
+						return
+					};
+					best_block_hash = next;
+				}
```

Comment:
Random thought: this "fetch the last available item from stream" could perhaps be wrapped into a little helper function and then that could be tested (though offhand I wouldn't know what to call it!)

Reply:
I'd just go for a single function that mutably borrows a stream and returns the last "next available thing" back, I reckon; no need for a trait :)